engine/execution/computation/computer/computer.go

```diff
@@ -276,96 +288,111 @@ func (e *blockComputer) executeSystemCollection(
-		e.log.Error().
-			Str("error_message", systemChunkTxResult.ErrorMessage).
-			Hex("block_id", logging.Entity(systemChunkCtx.BlockHeader)).
-			Bool("system_chunk_error", true).
-			Bool("critical_error", true).
-			Msg("error executing system chunk transaction")
-	}
-
-	snapshot := collectionView.(*delta.View).Interactions()
-	res.AddStateSnapshot(snapshot)
-	res.UpdateTransactionResultIndex(1)
-
-	compUsed, memUsed := res.ChunkComputationAndMemoryUsed(collectionIndex)
-	eventCounts, eventSize := res.ChunkEventCountsAndSize(collectionIndex)
-	e.metrics.ExecutionCollectionExecuted(time.Since(startedAt),
-		compUsed, memUsed,
-		eventCounts, eventSize,
-		snapshot.NumberOfRegistersTouched,
-		snapshot.NumberOfBytesWrittenToRegisters,
-		1,
-	)
-
 	return &flow.Collection{
-		Transactions: []*flow.TransactionBody{tx},
+		Transactions: collection.Transactions,
 	}, err
 }
 
 func (e *blockComputer) executeCollection(
 	blockSpan otelTrace.Span,
 	collectionIndex int,
 	txIndex uint32,
-	blockCtx fvm.Context,
 	collectionView state.View,
-	collection *entity.CompleteCollection,
+	collection collectionItem,
 	res *execution.ComputationResult,
 ) (uint32, error) {
 
-	e.log.Debug().
-		Hex("block_id", logging.Entity(blockCtx.BlockHeader)).
-		Hex("collection_id", logging.Entity(collection.Guarantee)).
-		Msg("executing collection")
-
 	// call tracing
 	startedAt := time.Now()
-	colSpan := e.tracer.StartSpanFromParent(blockSpan, trace.EXEComputeCollection)
-	defer func() {
+
+	var colSpan otelTrace.Span
+	if collection.isSystemCollection {
+		colSpan = e.tracer.StartSpanFromParent(
+			blockSpan,
+			trace.EXEComputeSystemCollection)
+
+		e.log.Debug().
+			Hex("block_id", logging.Entity(collection.ctx.BlockHeader)).
+			Msg("executing system collection")
+	} else {
+		colSpan = e.tracer.StartSpanFromParent(
+			blockSpan,
+			trace.EXEComputeCollection)
+
 		colSpan.SetAttributes(
 			attribute.Int("collection.txCount", len(collection.Transactions)),
-			attribute.String("collection.hash", collection.Guarantee.CollectionID.String()),
-		)
-		colSpan.End()
-	}()
+			attribute.String(
+				"collection.hash",
+				collection.Guarantee.CollectionID.String()))
+
+		e.log.Debug().
+			Hex("block_id", logging.Entity(collection.ctx.BlockHeader)).
+			Hex("collection_id", logging.Entity(collection.Guarantee)).
+			Msg("executing collection")
+	}
+
+	defer colSpan.End()
 
-	txCtx := fvm.NewContextFromParent(blockCtx, fvm.WithMetricsReporter(e.metrics), fvm.WithTracer(e.tracer))
 	for _, txBody := range collection.Transactions {
-		err := e.executeTransaction(txBody, colSpan, collectionView, txCtx, collectionIndex, txIndex, res, false)
+		err := e.executeTransaction(txBody, colSpan, collectionView, collection.ctx, collectionIndex, txIndex, res, false)
 		txIndex++
 		if err != nil {
 			return txIndex, err
 		}
 	}
+
+	if collection.isSystemCollection {
+		systemChunkTxResult := res.TransactionResults[len(res.TransactionResults)-1]
```

Comment:
do we need to be defensive here and verify that len is always == 1 here?

Reply:
we can't check for len == 1 since TransactionResults holds results for the entire block